Add the ability to view and list of leases metadata

chrishoffman opened the pull request:
This adds the ability to view and list lease metadata.  There are currently no methods to view lease metadata or to list leases by a given prefix.  Since there are methods that exist to revoke individual or a prefix of leases and renew individual leases, it makes sense to have endpoints that allow you to view this data as well.  The data that will be returned will only be metadata about the lease and will not return any internal lease data.

jefferai commented on vault/logical_system.go:
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
+		return handleError(err)
+	}
+
+	return &logical.Response{
```
Check `logical.ListResponse`, which does this for you!

chrishoffman replied:
Magic :).

vishalnayak commented on vault/logical_system.go:
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
+func (b *SystemBackend) handleLease(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	leaseID := data.Get("lease_id").(string)
+
```
Can we have a check for `leaseID == ""` here?

chrishoffman replied:
Will do.

vishalnayak commented on vault/logical_system.go:
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
+		Data: map[string]interface{}{
+			"lease_id":          leaseID,
+			"issue_time":        leaseTimes.IssueTime,
+			"expire_time":       nil,
```
Is there a reason for setting these 2 fields explicitly to `nil`?

chrishoffman replied:
This was really just to serialize to `null` in the serializer.  I guess omitting would probably be fine.

vishalnayak commented on vault/logical_system_test.go:
```diff
@@ -315,6 +317,179 @@ func TestSystemBackend_remount_system(t *testing.T) {
+	}
+
+	// Read a key with a LeaseID
+	req = logical.TestRequest(t, logical.ReadOperation, "secret/foo")
```
Generic secret backend does not attach a lease to secrets. I'm not sure if this test would work.

chrishoffman replied:
It does and I was surprised too and this how all the renew and revoke functions are written.  I think it might be something in the test harness.  Otherwise it isn't straightforward creating a non-token lease without some backing dynamic backend.  Maybe `pki` with `generate_lease` set `true`.

jefferai replied:
When we removed leases from the generic backend in the normal case we still kept the ability to have it issue leases specifically to make testing easier!

vishalnayak commented on vault/logical_system_test.go:
```diff
@@ -315,6 +317,179 @@ func TestSystemBackend_remount_system(t *testing.T) {
+	}
+
+	// Read a key with a LeaseID
+	req = logical.TestRequest(t, logical.ReadOperation, "secret/foo")
```
Same with this test as well.

briankassouf commented on vault/logical_system.go:
```diff
@@ -299,6 +300,42 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
+			},
+
+			&framework.Path{
+				Pattern: "lease" + framework.OptionalParamRegex("prefix"),
```
Instead of making this optional what if the two framework.Paths were merged with both a read op and a list op?

chrishoffman replied:
My only concern here was some arbitrary path would not be an "unsupported operation" on update operations because of the optional `prefix`.

briankassouf commented on vault/logical_system.go:
```diff
@@ -1270,6 +1307,66 @@ func (b *SystemBackend) handleTuneWriteCommon(
+
+func (b *SystemBackend) handleLeaseList(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	prefix := data.Get("prefix").(string)
```
Can you add a check to make sure prefix is not `""`

chrishoffman replied:
Empty prefix is valid and will list the top level keys.

briankassouf commented on vault/logical_system.go:
```diff
@@ -63,6 +63,7 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 				"replication/reindex",
 				"rotate",
 				"config/auditing/*",
+				"lease*",
```
Also you'll need to update the root paths test after changing this list

chrishoffman replied:
Thanks for the head up.  I wasn't aware of that test.

chrishoffman commented on vault/logical_system.go:
```diff
@@ -299,6 +300,42 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
+				},
+
+				Callbacks: map[logical.Operation]framework.OperationFunc{
+					logical.UpdateOperation: b.handleLease,
```
Since `lease_id` is sensitive, we determined it made sense to pass `lease_id` as a parameter with an update operation.  This is currently the preferred method to use in `/sys/renew` and `/sys/revoke`.

vishalnayak commented on vault/logical_system.go:
```diff
@@ -347,7 +371,7 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 
 			&framework.Path{
-				Pattern: "revoke-force/(?P<prefix>.+)",
+				Pattern: "(lease/)?revoke-force" + framework.OptionalParamRegex("prefix"),
```
Before this change, I think the `prefix` parameter was mandatory and would always be set. If it was not supplied, parsing the path pattern would error out. Now that it is an optional param, missing `prefix` I think will not error out. We might want to check explicitly that the required parameters (but optionally set in the URL) are in fact set. This comment applies to all the similar changes below.

chrishoffman replied:
You are right.  I got a little ambitious in my refactoring.  I'll revert out these changes.

vishalnayak commented on vault/logical_system.go:
```diff
@@ -55,14 +55,16 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 				"rotate",
 				"config/auditing/*",
+				"lease/lookup*",
+				"lease/revoke-prefix/*",
```
Can we also add `revoke-force/*` and `lease/revoke-force/*` to this list?

chrishoffman replied:
It is probably the right thing to do but this would be a change in behavior.  I just want to make sure that what we want to do.  I will note it in the changelog if we decide to go that route.

vishalnayak replied:
If `revoke-prefix` is a privileged operation, `revoke-force` must be I suppose. If we decide to do this, we should document it in CL as you said, in the docs and as a breaking behavior in the upgrade notes.

jefferai replied:
Generally we reserve root paths for functions that can be super destructive or super sensitive from a security standpoint. `revoke-force` does fit that bill and probably should be privileged, but I wonder if we should reserve that change for 0.8; people will be more likely to expect breaking changes for that release and I don't think this is super high priority.

vishalnayak commented on vault/logical_system.go:
```diff
@@ -1274,6 +1300,56 @@ func (b *SystemBackend) handleTuneWriteCommon(
+			"renewable":     leaseTimes.renewable(),
+		},
+	}
+	if !leaseTimes.LastRenewalTime.IsZero() {
```
Sorry for circling around on this. On a second thought, the idea of having fields with default values (like you were doing earlier) does seem like a sensible thing to do too. What might be of concern is that this API will not have deterministic number of response fields. That is okay I guess but I am not sure if that is a problem. Bringing this up so we can all agree on this and move forward.

vishalnayak commented on vault/logical_system.go:
```diff
@@ -2429,4 +2505,19 @@ This path responds to the following HTTP methods.
 	},
+
+	"lease": {
+		``,
```
Reminder to fill these out.

jefferai commented on vault/logical_system.go:
```diff
@@ -55,14 +55,16 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 				"replication/reindex",
 				"rotate",
 				"config/auditing/*",
+				"lease/lookup*",
```
Can this be `leases` not `lease`? It's a collection so plural makes sense. (As opposed to a singular system like `replication`, e.g. if it was `lease-management`, but that's too wordy.)

chrishoffman replied:
I debated this but wanted to match the other paths that had similar functions, like `token` and `policy`.  I am not strongly either way and was just trying for consistency.

jefferai replied:
The difference with `policy` is that it's supposed to be you acting on a single one, e.g. PUT (write) a policy, GET a policy, and so on. If there were a collection of policy-related APIs under a prefix, having that prefix be `policies` instead of `policy` would be the right move IMHO.

jefferai replied:
Also, lease lookup should definitely not be a root path.

jefferai commented on vault/expiration.go:
```diff
@@ -841,11 +841,18 @@ type leaseEntry struct {
+	return false
+}
+
+func (le *leaseEntry) renewableErr() error {
```
What is the point of doing this?

chrishoffman replied:
I thought this was a bit confusing and was just consolidating the logic for dealing with the error, which the expiration manager uses, and returning the value of renewable.

jefferai replied:
I'm concerned about people defaulting to use this function when in fact the exact error returned can be useful, e.g. in a debugging/logs context.

jefferai commented on vault/logical_system.go:
```diff
@@ -1274,6 +1300,56 @@ func (b *SystemBackend) handleTuneWriteCommon(
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"id":            leaseID,
+			"creation_time": leaseTimes.IssueTime,
```
This has a different type than `creation_time` in the token store; there, you set this to `issue_time`.

jefferai replied:
There's a type discrepancy with `last_renewal_time` too.

chrishoffman commented:
This should be ready for review again.  

Notable changes since last full review:
* Added documentation
* Nested "lease" related paths under `/sys/leases`.  Backwards compatibility is maintained.
* Token lookup and lease lookup fields merged

I ended up moving a lot of documentation around.  Should we provide landing pages for the old docs locations that point to the new locations?

briankassouf commented on vault/logical_system.go:
```diff
@@ -325,7 +363,7 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
 
 			&framework.Path{
-				Pattern: "revoke" + framework.OptionalParamRegex("url_lease_id"),
+				Pattern: "(leases/)?revoke" + framework.OptionalParamRegex("url_lease_id"),
```
👍

vishalnayak commented on vault/logical_system.go:
```diff
@@ -1274,6 +1314,61 @@ func (b *SystemBackend) handleTuneWriteCommon(
 	return nil, nil
 }
 
+// handleLease is use to view the metadata for a given LeaseID
```
s/handleLease/handleLeaseLookup

vishalnayak commented on vault/logical_system.go:
```diff
@@ -1274,6 +1314,61 @@ func (b *SystemBackend) handleTuneWriteCommon(
+	if leaseTimes == nil {
+		return logical.ErrorResponse("invalid lease"), logical.ErrInvalidRequest
+	}
+
```
Can we add a comment here explaining why we are setting values to nil beforehand. This will avoid people removing it in future.

briankassouf commented on vault/logical_system.go:
```diff
@@ -299,7 +302,43 @@ func NewSystemBackend(core *Core, config *logical.BackendConfig) (logical.Backen
+			},
+
+			&framework.Path{
+				Pattern: "leases/lookup",
```
Do we want this to be a root protected path?

jefferai replied:
@briankassouf Since `lease/lookup` doesn't return sensitive information (e.g. the contents of internal data) and you have to know the lease ID ahead of time I don't think it needs to be root. In fact, I am actually thinking we may want to add it to the `default` policy. _Listing_ on the other hand *does* divulge lease IDs which can be used to e.g. renew them, so that should be root protected.

jefferai commented:
@vishalnayak it's not a breaking change; `sys/revoke-force` remains non-root (for now), but `sys/leases/revoke-force` is root. Since that's what documentation will focus on we'll get people used to using that, then eventually can change over the older path.
